### DISCUSSION TITLE
Can't access NuxtApp after await

### DISCUSSION THREAD

**Discussion by rikbrowning** [nuxt/nuxt#17496] [Questions]
I created the following example: https://stackblitz.com/edit/nuxt-starter-w8fr5y?file=composables%2FuseUsers.ts

Inside my composable the useFetch now fails because it can no longer access the NuxtApp. Is this a limitation? How should we handle having multiple async in a composable but still wanting to have access to NuxtApp? Note in my example it fails for useFetch but it also would fail for useState, basically any nuxt functionality that needs NuxtApp

**Comment by wobsoriano** [comment]
`useNuxtApp` only works on setup function and plugins. To fix that you can check the process

```ts
import { useNuxtApp } from '#app';

export default async function useUsers() {
  const app = useNuxtApp();
  if (process.client) {
    await pretendAuth();
  }
  if (process.server) {
    // console.log('app', app); ERROR
  }
  return useFetch('/api/users');
}
```

**Reply by rikbrowning** [reply]
So how outside of those things are you meant to gain access to data being exposed via plugins? Such as auth data?

**Reply by wobsoriano** [reply]
Try useState and page middlewares

**Reply by rikbrowning** [reply]
But to use useState I need access to the NuxtApp which I cannot access after a single await statement? In order to useState I would have to declare that and have it set prior to any async event to get data. What are the different limitations we have around the exposed composables? It was just the other day I discovered a comment detailing restrictions on useFetch.

**Reply by wobsoriano** [reply]
If you are trying to fetch data on server and set the useState value, you can create a plugin for that

```ts
// ~/composables/useAuthToken.ts
export default function useAuthToken() {
  return useState('auth_token')
}
```

```ts
// ~/plugins/someplugin.server.ts
export default defineNuxtPlugin(async () => {
  const token = useAuthToken()

  const data = await $fetch('/api/get-auth')

  token.value = data.token
})
```

```ts
// ~/middleware/auth.ts
export default defineNuxtRouteMiddleware((to, _from) => {
  const token = useAuthToken()
  if (!token.value) {
    return navigateTo('/login')
  }
})
```

**Reply by rikbrowning** [reply]
Yeah I have plugins loading the auth details on both client and server plugins. I guess auth was a bad example, was more asking what is the best practice for data access in composables, especially when we could have the scenario of a composable using a composable right? Any async behaviour could break access to useState.

**Reply by wobsoriano** [reply]
Yeah you will have to do something like

```ts
export function useSomething() {
  const state= useState('user', () => null)

  const fetchUser = async () => {
    const user = await someFunc();
    state.value = user
  }

  fetchUser()

  return state
}
```
